### PR TITLE
DM-46813: Add sequence_to_range_str to compactify lists

### PR DESCRIPTION
**Pull request by mfisherlevine**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/utils/iteration.py [hunk 1]
```diff
@@ -107,3 +107,195 @@ def isplit(string: T, sep: T) -> Iterator[T]:
...
+    if isinstance(val0, str) and isinstance(val1, str):
+        prefix0, num_suffix0 = _extract_numeric_suffix(val0)
+        prefix1, num_suffix1 = _extract_numeric_suffix(val1)
+        if prefix0 == prefix1 and num_suffix0 is not None and num_suffix1 is not None:
+            if num_suffix0 == num_suffix1:
+                dvn = val0
+            else:
+                dvn = f"{val0}..{val1}"
+                if stride > 1:
+                    dvn += f":{stride}"
```
Comment (RobertLuptonTheGood) [inline python/lsst/utils/iteration.py:173]: I'd add a test case for non-unit strides.  They are common in HSC. `[10, 11, 12]` should trigger it

File: python/lsst/utils/iteration.py [hunk 2]
```diff
@@ -107,3 +107,195 @@ def isplit(string: T, sep: T) -> Iterator[T]:
...
+            dvn = sval0
+        elif isinstance(val0, int) and isinstance(val1, int):
+            if val1 == val0 + stride:
+                dvn = f"{sval0}^{sval1}"
+            else:
+                dvn = f"{sval0}..{sval1}"
+                if stride > 1:
+                    dvn += f":{stride}"
+        else:
+            dvn = f"{sval0}^{sval1}"
```
Comment (RobertLuptonTheGood) [inline python/lsst/utils/iteration.py:189]: This is also a common case.  Even `[10, 100]` should trigger it;  so I'm confused

File: python/lsst/utils/iteration.py [hunk 3]
```diff
@@ -107,3 +107,195 @@ def isplit(string: T, sep: T) -> Iterator[T]:
...
+    if len(values) > 1 and _is_list_of_ints(values):
+        differences = [values[i + 1] - values[i] for i in range(len(values) - 1)]
+        stride = min(differences) if differences else 1
+        stride = max(stride, 1)
+
+    val_name: list[str] = []
+    val0 = values[0]
+    val1 = val0
+    for val in values[1:]:
+        if type(val) is not type(val1):
```
Comment (RobertLuptonTheGood) [inline python/lsst/utils/iteration.py:269]: Is this live?  I thought you said that you now refused to mixed input types, so is this even possible?

**Review by RobertLuptonTheGood — APPROVED** [review]
Looks OK, but I don't understand how the code coverage failed to cover some apparently common cases.  I'd check the logic.

File: python/lsst/utils/iteration.py (hunk 3 above)
Comment (mfisherlevine) [reply python/lsst/utils/iteration.py:269]: Agreed - removed.

**Pull request merged**